README.md
[![view on npm](https://img.shields.io/npm/v/@uttori/plugin-auth-simple.svg)](https://www.npmjs.com/package/@uttori/plugin-auth-simple)
[![npm module downloads](https://img.shields.io/npm/dt/@uttori/plugin-auth-simple.svg)](https://www.npmjs.com/package/@uttori/plugin-auth-simple)
[![Build Status](https://travis-ci.com/uttori/uttori-plugin-auth-simple.svg?branch=master)](https://travis-ci.com/uttori/uttori-plugin-auth-simple)
[![Coverage Status](https://coveralls.io/repos/github/uttori/uttori-plugin-auth-simple/badge.svg?branch=master)](https://coveralls.io/github/uttori/uttori-plugin-auth-simple?branch=master)
![npm bundle size](https://img.shields.io/bundlephobia/minzip/@uttori/plugin-auth-simple?label=Minified%20%2B%20GZip)

# Uttori Plugin - Auth Simple

A plugin to add very simple authentication using [express-session](https://github.com/expressjs/session). A great starting point for any Uttori auth solution.

On `/login` success, as HTML it will redirect to the provided `loginRedirectPath` with a `302` status code.

On `/login` success, as JSON it will return the payload from `validateLogin(request.body)` and set it on the session as `request.session.profile`.

On `/login` error, as HTML it will redirect to the provided `loginPath` with a `401` status code.

On `/login` error, as JSON it will return the payload `{ error: true }` with a `401` status code.

Note: When using with browser based `fetch` or the like, ensure `withCredentials` is set to true.

## Install

```bash
npm install --save uttori-plugin-auth-simple
```

## Config

```js
{
    // Registration Events
    events: {
      bindRoutes: ['bind-routes'],
      validateConfig: ['validate-config'],
    },

    // Login Path
    loginPath: '/login',

    // Login Redirect Path
    loginRedirectPath: '/',

    // Login Route Middleware
    loginMiddleware: [],

    // Logout Path
    logoutPath: '/logout',

    // Logout Redirect Path
    logoutRedirectPath: '/',

    // Logout Route Middleware
    logoutMiddleware: [],

    // Validation function that will recieve the request and returns an object to be used as the session payload.
    // If the session is invalid it should return false.
    validateLogin: (request) => {},
}
```

* * *

## API Reference

## Classes

<dl>
<dt><a href="#AuthSimple">AuthSimple</a></dt>
<dd><p>Uttori Auth (Simple)</p>
</dd>
</dl>

## Functions

<dl>
<dt><a href="#asyncHandler">asyncHandler()</a> : <code>function</code></dt>
<dd></dd>
</dl>

## Typedefs

<dl>
<dt><a href="#AuthSimpleConfigEvents">AuthSimpleConfigEvents</a> : <code>object</code></dt>
<dd></dd>
<dt><a href="#AuthSimpleConfig">AuthSimpleConfig</a> : <code>object</code></dt>
<dd></dd>
</dl>

<a name="AuthSimple"></a>

## AuthSimple
Uttori Auth (Simple)

**Kind**: global class  

* [AuthSimple](#AuthSimple)
    * [.configKey](#AuthSimple.configKey) ⇒ <code>string</code>
    * [.defaultConfig()](#AuthSimple.defaultConfig) ⇒ [<code>AuthSimpleConfig</code>](#AuthSimpleConfig)
    * [.validateConfig(config, _context)](#AuthSimple.validateConfig)
    * [.register(context)](#AuthSimple.register)
    * [.bindRoutes(server, context)](#AuthSimple.bindRoutes)
    * [.login(context)](#AuthSimple.login) ⇒ <code>function</code>
    * [.logout(context)](#AuthSimple.logout) ⇒ <code>function</code>

<a name="AuthSimple.configKey"></a>

### AuthSimple.configKey ⇒ <code>string</code>
The configuration key for plugin to look for in the provided configuration.

**Kind**: static property of [<code>AuthSimple</code>](#AuthSimple)  
**Returns**: <code>string</code> - The configuration key.  
**Example** *(AuthSimple.configKey)*  
```js
const config = { ...AuthSimple.defaultConfig(), ...context.config[AuthSimple.configKey] };
```
<a name="AuthSimple.defaultConfig"></a>

### AuthSimple.defaultConfig() ⇒ [<code>AuthSimpleConfig</code>](#AuthSimpleConfig)
The default configuration.

**Kind**: static method of [<code>AuthSimple</code>](#AuthSimple)  
**Returns**: [<code>AuthSimpleConfig</code>](#AuthSimpleConfig) - The configuration.  
**Example** *(AuthSimple.defaultConfig())*  
```js
const config = { ...AuthSimple.defaultConfig(), ...context.config[AuthSimple.configKey] };
```
<a name="AuthSimple.validateConfig"></a>

### AuthSimple.validateConfig(config, _context)
Validates the provided configuration for required entries.

**Kind**: static method of [<code>AuthSimple</code>](#AuthSimple)  

| Param | Type | Description |
| --- | --- | --- |
| config | <code>object</code> | A configuration object. |
| config.configKey | <code>object</code> | A configuration object specifically for this plugin. |
| _context | <code>object</code> | Unused. |

**Example** *(AuthSimple.validateConfig(config, _context))*  
```js
AuthSimple.validateConfig({ ... });
```
<a name="AuthSimple.register"></a>

### AuthSimple.register(context)
Register the plugin with a provided set of events on a provided Hook system.

**Kind**: static method of [<code>AuthSimple</code>](#AuthSimple)  

| Param | Type | Description |
| --- | --- | --- |
| context | <code>object</code> | A Uttori-like context. |
| context.hooks | <code>object</code> | An event system / hook system to use. |
| context.hooks.on | <code>function</code> | An event registration function. |
| context.config | <code>object</code> | A provided configuration to use. |
| context.config.events | <code>object</code> | An object whose keys correspong to methods, and contents are events to listen for. |

**Example** *(AuthSimple.register(context))*  
```js
const context = {
  hooks: {
    on: (event, callback) => { ... },
  },
  config: {
    [AuthSimple.configKey]: {
      ...,
      events: {
        bindRoutes: ['bind-routes'],
      },
    },
  },
};
AuthSimple.register(context);
```
<a name="AuthSimple.bindRoutes"></a>

### AuthSimple.bindRoutes(server, context)
Add the login & logout routes to the server object.

**Kind**: static method of [<code>AuthSimple</code>](#AuthSimple)  

| Param | Type | Description |
| --- | --- | --- |
| server | <code>object</code> | An Express server instance. |
| server.post | <code>function</code> | Function to register route. |
| context | <code>object</code> | A Uttori-like context. |
| context.config | <code>object</code> | A provided configuration to use. |

**Example** *(AuthSimple.bindRoutes(server, context))*  
```js
const context = {
  hooks: {
    on: (event, callback) => { ... },
  },
  config: {
    [AuthSimple.configKey]: {
      loginPath: '/login',
      logoutPath: '/logout',
      loginMiddleware: [ ... ],
      logoutMiddleware: [ ... ],
    },
  },
};
AuthSimple.bindRoutes(server, context);
```
<a name="AuthSimple.login"></a>

### AuthSimple.login(context) ⇒ <code>function</code>
The Express route method to process the login request and provide a response or redirect.

**Kind**: static method of [<code>AuthSimple</code>](#AuthSimple)  
**Returns**: <code>function</code> - - The function to pass to Express.  

| Param | Type | Description |
| --- | --- | --- |
| context | <code>object</code> | A Uttori-like context. |
| context.config | <code>object</code> | A provided configuration to use. |

**Example** *(AuthSimple.login(context)(request, response, next))*  
```js
server.post('/login', AuthSimple.login(context));
```
<a name="AuthSimple.logout"></a>

### AuthSimple.logout(context) ⇒ <code>function</code>
The Express route method to process the logout request and clear the session.

**Kind**: static method of [<code>AuthSimple</code>](#AuthSimple)  
**Returns**: <code>function</code> - - The function to pass to Express.  

| Param | Type | Description |
| --- | --- | --- |
| context | <code>object</code> | A Uttori-like context. |
| context.config | <code>object</code> | A provided configuration to use. |

**Example** *(AuthSimple.login(context)(request, response, _next))*  
```js
server.post('/logout', AuthSimple.login(context));
```
<a name="asyncHandler"></a>

## asyncHandler() : <code>function</code>
**Kind**: global function  
<a name="AuthSimpleConfigEvents"></a>

## AuthSimpleConfigEvents : <code>object</code>
**Kind**: global typedef  
**Properties**

| Name | Type | Description |
| --- | --- | --- |
| bindRoutes | <code>Array.&lt;string&gt;</code> | The collection of events to bind to for setting up the login and logout routes. |
| validateConfig | <code>Array.&lt;string&gt;</code> | The collection of events to bind to for validating the configuration. |

<a name="AuthSimpleConfig"></a>

## AuthSimpleConfig : <code>object</code>
**Kind**: global typedef  
**Properties**

| Name | Type | Default | Description |
| --- | --- | --- | --- |
| [events] | [<code>AuthSimpleConfigEvents</code>](#AuthSimpleConfigEvents) |  | Events to bind to. |
| [loginPath] | <code>string</code> | <code>&quot;&#x27;/login&#x27;&quot;</code> | The path to the login endpoint. |
| [loginRedirectPath] | <code>string</code> | <code>&quot;&#x27;/&#x27;&quot;</code> | The path to redirect to after logging in. |
| [loginMiddleware] | <code>Array.&lt;function()&gt;</code> | <code>[]</code> | The middleware to use on the login route. |
| [logoutPath] | <code>string</code> | <code>&quot;&#x27;/logout&#x27;&quot;</code> | The path to the logout endpoint. |
| [logoutRedirectPath] | <code>string</code> | <code>&quot;&#x27;/&#x27;&quot;</code> | The path to redirect to after logging out. |
| [logoutMiddleware] | <code>Array.&lt;function()&gt;</code> | <code>[]</code> | The middleware to use on the logout route. |
| validateLogin | <code>function</code> |  | A function that accepts a Request object and validates a users credentials, if they are invalid it should return false. |


* * *

## Tests

To run the test suite, first install the dependencies, then run `npm test`:

```bash
npm install
npm test
DEBUG=Uttori* npm test
```

## Contributors

* [Matthew Callis](https://github.com/MatthewCallis)

## License

* [MIT](LICENSE)


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. This project adheres to [Semantic Versioning](http://semver.org/).

## [1.0.1](https://github.com/uttori/uttori-plugin-auth-simple/compare/v1.0.0...v1.0.1) - 2023-04-13

- 🐛 Fix inverted login in `register`

## [1.0.0](https://github.com/uttori/uttori-plugin-auth-simple/compare/v1.0.0...v1.0.0) - 2023-04-13

- 🧰 Released


test/index.test.js
// @ts-nocheck
const test = require('ava');
const sinon = require('sinon');
const request = require('supertest');
const AuthSimple = require('../src');

test('AuthSimple.register(context): can register', (t) => {
  t.notThrows(() => {
    AuthSimple.register({ hooks: { on: () => {} }, config: { [AuthSimple.configKey]: { events: { callback: [] } } } });
  });
});

test('AuthSimple.register(context): errors without event dispatcher', (t) => {
  t.throws(() => {
    AuthSimple.register({ hooks: {} });
  }, { message: 'Missing event dispatcher in \'context.hooks.on(event, callback)\' format.' });
});

test('AuthSimple.register(context): errors without events', (t) => {
  t.throws(() => {
    AuthSimple.register({ hooks: { on: () => {} }, config: { [AuthSimple.configKey]: { } } });
  }, { message: 'Missing events to listen to for in \'config.events\'.' });
});

test('AuthSimple.register(context): does not error with events corresponding to missing methods', (t) => {
  t.notThrows(() => {
    AuthSimple.register({
      hooks: {
        on: () => {},
      },
      config: {
        [AuthSimple.configKey]: {
          events: {
            test: ['test'],
            bindRoutes: ['bind-routes'],
            validateConfig: ['validate-config'],
          },
        },
      },
    });
  });
});

test('AuthSimple.defaultConfig(): can return a default config', (t) => {
  t.notThrows(AuthSimple.defaultConfig);
});

test('AuthSimple.validateConfig(config, _context): throws when configuration key is missing', (t) => {
  t.throws(() => {
    AuthSimple.validateConfig({});
  }, { message: 'Config Error: \'uttori-plugin-auth-simple\' configuration key is missing.' });
});

test('AuthSimple.validateConfig(config, _context): throws when loginPath is not a string', (t) => {
  t.throws(() => {
    AuthSimple.validateConfig({
      [AuthSimple.configKey]: {
        loginPath: 10,
      },
    });
  }, { message: 'Config Error: `loginPath` should be a string server route to where credentials should be POSTed to.' });
});

test('AuthSimple.validateConfig(config, _context): throws when loginRedirectPath is not a string', (t) => {
  t.throws(() => {
    AuthSimple.validateConfig({
      [AuthSimple.configKey]: {
        loginPath: '/login',
        loginRedirectPath: false,
      },
    });
  }, { message: 'Config Error: `loginRedirectPath` should be a string server route to where a successful login should navigate to.' });
});

test('AuthSimple.validateConfig(config, _context): throws when loginMiddleware is not an array', (t) => {
  t.throws(() => {
    AuthSimple.validateConfig({
      [AuthSimple.configKey]: {
        loginPath: '/login',
        loginRedirectPath: '/',
        loginMiddleware: {},
      },
    });
  }, { message: 'Config Error: `loginMiddleware` should be an array of middlewares or an empty array.' });
});

test('AuthSimple.validateConfig(config, _context): throws when logoutPath is not a string', (t) => {
  t.throws(() => {
    AuthSimple.validateConfig({
      [AuthSimple.configKey]: {
        loginPath: '/login',
        loginRedirectPath: '/',
        loginMiddleware: [],
        logoutPath: true,
      },
    });
  }, { message: 'Config Error: `logoutPath` should be a string server route to where logout requests should be POSTed to.' });
});

test('AuthSimple.validateConfig(config, _context): throws when logoutRedirectPath is not a string', (t) => {
  t.throws(() => {
    AuthSimple.validateConfig({
      [AuthSimple.configKey]: {
        loginPath: '/login',
        loginRedirectPath: '/',
        loginMiddleware: [],
        logoutPath: '/logout',
        logoutRedirectPath: true,
      },
    });
  }, { message: 'Config Error: `logoutRedirectPath` should be a string server route to where a logout should navigate to.' });
});

test('AuthSimple.validateConfig(config, _context): throws when logoutMiddleware is not an array', (t) => {
  t.throws(() => {
    AuthSimple.validateConfig({
      [AuthSimple.configKey]: {
        loginPath: '/login',
        loginRedirectPath: '/',
        loginMiddleware: [],
        logoutPath: '/logout',
        logoutRedirectPath: '/',
        logoutMiddleware: {},
      },
    });
  }, { message: 'Config Error: `logoutMiddleware` should be an array of middlewares or an empty array.' });
});

test('AuthSimple.validateConfig(config, _context): throws when validateLogin is not a function', (t) => {
  t.throws(() => {
    AuthSimple.validateConfig({
      [AuthSimple.configKey]: {
        loginPath: '/login',
        loginRedirectPath: '/',
        loginMiddleware: [],
        logoutPath: '/logout',
        logoutRedirectPath: '/',
        logoutMiddleware: [],
        validateLogin: '',
      },
    });
  }, { message: 'Config Error: `validateLogin` should be a function to validate the request body.' });
});

test('AuthSimple.validateConfig(config, _context):can validate', (t) => {
  t.notThrows(() => {
    AuthSimple.validateConfig({
      [AuthSimple.configKey]: {
        loginPath: '/login',
        loginRedirectPath: '/',
        loginMiddleware: [],
        logoutPath: '/logout',
        logoutRedirectPath: '/',
        logoutMiddleware: [],
        validateLogin: () => {},
      },
    });
  });
});

test('AuthSimple.bindRoutes(server, context): can bind routes', (t) => {
  const post = sinon.spy();
  const server = { post };
  AuthSimple.bindRoutes(server, {
    config: {
      [AuthSimple.configKey]: {
        events: {
          bindRoutes: ['bind-routes'],
          validateConfig: ['validate-config'],
        },
        loginPath: '/login',
        loginRedirectPath: '/',
        loginMiddleware: [],
        logoutPath: '/logout',
        logoutRedirectPath: '/',
        logoutMiddleware: [],
        validateLogin: (_request) => true,
      },
    },
  });
  t.true(post.calledTwice);
});

test('AuthSimple.login(context): returns a Express route and will call next if validateLogin is missing', async (t) => {
  const { serverSetup } = require('./_helpers/server');
  const loginPath = '/login';
  const context = {
    config: {
      [AuthSimple.configKey]: {
        events: {
          bindRoutes: ['bind-routes'],
          validateConfig: ['validate-config'],
        },
        loginPath,
        loginRedirectPath: '/1',
        loginMiddleware: [],
        logoutPath: '/logout',
        logoutRedirectPath: '/2',
        logoutMiddleware: [],
        validateLogin: false,
      },
    },
  };
  const server = serverSetup();
  AuthSimple.bindRoutes(server, context);
  const response = await request(server).post(loginPath).field('token', 'abcd');
  t.is(response.status, 404);
});

test('AuthSimple.login(context): returns a Express route and can attempt to login as HTML', async (t) => {
  t.plan(2);
  const { serverSetup } = require('./_helpers/server');
  const loginPath = '/login';
  const context = {
    config: {
      [AuthSimple.configKey]: {
        events: {
          bindRoutes: ['bind-routes'],
          validateConfig: ['validate-config'],
        },
        loginPath,
        loginRedirectPath: '/1',
        loginMiddleware: [],
        logoutPath: '/logout',
        logoutRedirectPath: '/2',
        logoutMiddleware: [],
        validateLogin: (req) => {
          if (req?.body?.token === 'abcd') {
            return { admin: true };
          }
          return false;
        },
      },
    },
  };
  const server = serverSetup();
  AuthSimple.bindRoutes(server, context);
  const response = await request(server).post(loginPath).field('token', 'abcd');
  t.is(response.status, 302);
  t.is(response.text, 'Found. Redirecting to /login');
});

test('AuthSimple.login(context): returns a Express route and can attempt to logout as HTML', async (t) => {
  t.plan(2);
  const { serverSetup } = require('./_helpers/server');
  const loginPath = '/login';
  const logoutPath = '/logout';
  const context = {
    config: {
      [AuthSimple.configKey]: {
        events: {
          bindRoutes: ['bind-routes'],
          validateConfig: ['validate-config'],
        },
        loginPath,
        loginRedirectPath: '/1',
        loginMiddleware: [],
        logoutPath,
        logoutRedirectPath: '/2',
        logoutMiddleware: [],
        validateLogin: (req) => {
          if (req?.body?.token === 'abcd') {
            return { admin: true };
          }
          return false;
        },
      },
    },
  };
  const server = serverSetup();
  AuthSimple.bindRoutes(server, context);
  const response = await request(server).post(logoutPath);
  t.is(response.status, 302);
  t.is(response.text, 'Found. Redirecting to /2');
});


test/_helpers/server.js
/* eslint-disable no-console */
// Server
const express = require('express');
const session = require('express-session');
const MemoryStore = require('memorystore')(session);

const serverSetup = () => {
  // Server Setup
  const server = express();
  server.set('port', process.env.PORT || 8123);
  server.set('ip', process.env.IP || '127.0.0.1');

  // Setup sessions.
  server.use(session({
    secret: 'auth-simple',
    name: 'session',
    resave: false,
    saveUninitialized: false,
    cookie: {
      maxAge: 60000,
      sameSite: true,
      secure: true,
    },
    store: new MemoryStore({
      checkPeriod: 3600000, // prune expired entries every hour
    }),
  }));

  // process.title (for stopping after)
  if (process.argv[2] && process.argv[2] !== 'undefined') {
    console.log('Setting process.title:', typeof process.argv[2], process.argv[2]);
    // eslint-disable-next-line prefer-destructuring
    process.title = process.argv[2];
  }

  // Is this a require()?
  if (require.main === module) {
    console.log('Starting test server...');
    server.listen(server.get('port'), server.get('ip'));
  }

  return server;
};

module.exports = { serverSetup };


types/index.d.ts
declare module '@uttori/plugin-auth-simple';

declare module 'index' {
    type AuthSimpleConfigEvents = {
        bindRoutes: string[];
        validateConfig: string[];
    };
    type AuthSimpleConfig = {
        events?: AuthSimpleConfigEvents;
        loginPath?: string;
        loginRedirectPath?: string;
        loginMiddleware?: Function[];
        logoutPath?: string;
        logoutRedirectPath?: string;
        logoutMiddleware?: Function[];
        validateLogin: Function;
    };
    export class AuthSimple {
      static get configKey(): string;

      static defaultConfig(): AuthSimpleConfig;

      static validateConfig(config: {
            configKey: object;
        }, _context: object): void;

      static register(context: {
            hooks: {
                on: Function;
            };
            config: {
                events: object;
            };
        }): void;

      static bindRoutes(server: {
            post: Function;
        }, context: {
            config: object;
        }): void;

      static login(context: {
            config: object;
        }): Function;

      static logout(context: {
            config: object;
        }): Function;
    }
    namespace AuthSimple {
        export { AuthSimpleConfigEvents, AuthSimpleConfig };
    }
}


src/index.js
const debug = require('debug')('Uttori.Plugin.AuthSimple');

/** @type {Function} */
const asyncHandler = (fn) => (request, response, next) => Promise.resolve(fn(request, response, next)).catch(next);

/**
 * @typedef AuthSimpleConfigEvents
 * @type {object}
 * @property {string[]} bindRoutes The collection of events to bind to for setting up the login and logout routes.
 * @property {string[]} validateConfig The collection of events to bind to for validating the configuration.
 */

/**
 * @typedef AuthSimpleConfig
 * The AuthSimple configuration object.
 * @type {object}
 * @property {AuthSimpleConfigEvents} [events] Events to bind to.
 * @property {string} [loginPath='/login'] The path to the login endpoint.
 * @property {string} [loginRedirectPath='/'] The path to redirect to after logging in.
 * @property {Function[]} [loginMiddleware=[]] The middleware to use on the login route.
 * @property {string} [logoutPath='/logout'] The path to the logout endpoint.
 * @property {string} [logoutRedirectPath='/'] The path to redirect to after logging out.
 * @property {Function[]} [logoutMiddleware=[]] The middleware to use on the logout route.
 * @property {Function} validateLogin A function that accepts a Request object and validates a users credentials, if they are invalid it should return false.
 */

/**
 * Uttori Auth (Simple)
 *
 * @example <caption>AuthSimple</caption>
 * const content = AuthSimple.storeFile(request);
 * @class
 */
class AuthSimple {
  /**
   * The configuration key for plugin to look for in the provided configuration.
   *
   * @type {string}
   * @returns {string} The configuration key.
   * @example <caption>AuthSimple.configKey</caption>
   * const config = { ...AuthSimple.defaultConfig(), ...context.config[AuthSimple.configKey] };
   * @static
   */
  static get configKey() {
    return 'uttori-plugin-auth-simple';
  }

  /**
   * The default configuration.
   *
   * @returns {AuthSimpleConfig} The configuration.
   * @example <caption>AuthSimple.defaultConfig()</caption>
   * const config = { ...AuthSimple.defaultConfig(), ...context.config[AuthSimple.configKey] };
   * @static
   */
  static defaultConfig() {
    return {
      // Login Path
      loginPath: '/login',

      // Login Redirect Path
      loginRedirectPath: '/',

      // Login Route Middleware
      loginMiddleware: [],

      // Logout Path
      logoutPath: '/logout',

      // Logout Redirect Path
      logoutRedirectPath: '/',

      // Logout Route Middleware
      logoutMiddleware: [],

      // Validation function that will recieve the request body that returns an object to be used as the session payload.
      validateLogin: () => {},
    };
  }

  /**
   * Validates the provided configuration for required entries.
   *
   * @param {object} config - A configuration object.
   * @param {object} config.configKey - A configuration object specifically for this plugin.
   * @param {object} _context - Unused.
   * @example <caption>AuthSimple.validateConfig(config, _context)</caption>
   * AuthSimple.validateConfig({ ... });
   * @static
   */
  static validateConfig(config, _context) {
    debug('Validating config...');
    if (!config || !config[AuthSimple.configKey]) {
      const error = `Config Error: '${AuthSimple.configKey}' configuration key is missing.`;
      debug(error);
      throw new Error(error);
    }
    if (typeof config[AuthSimple.configKey].loginPath !== 'string') {
      const error = 'Config Error: `loginPath` should be a string server route to where credentials should be POSTed to.';
      debug(error);
      throw new Error(error);
    }
    if (typeof config[AuthSimple.configKey].loginRedirectPath !== 'string') {
      const error = 'Config Error: `loginRedirectPath` should be a string server route to where a successful login should navigate to.';
      debug(error);
      throw new Error(error);
    }
    if (!Array.isArray(config[AuthSimple.configKey].loginMiddleware)) {
      const error = 'Config Error: `loginMiddleware` should be an array of middlewares or an empty array.';
      debug(error);
      throw new Error(error);
    }
    if (typeof config[AuthSimple.configKey].logoutPath !== 'string') {
      const error = 'Config Error: `logoutPath` should be a string server route to where logout requests should be POSTed to.';
      debug(error);
      throw new Error(error);
    }
    if (typeof config[AuthSimple.configKey].logoutRedirectPath !== 'string') {
      const error = 'Config Error: `logoutRedirectPath` should be a string server route to where a logout should navigate to.';
      debug(error);
      throw new Error(error);
    }
    if (!Array.isArray(config[AuthSimple.configKey].logoutMiddleware)) {
      const error = 'Config Error: `logoutMiddleware` should be an array of middlewares or an empty array.';
      debug(error);
      throw new Error(error);
    }
    if (typeof config[AuthSimple.configKey].validateLogin !== 'function') {
      const error = 'Config Error: `validateLogin` should be a function to validate the request body.';
      debug(error);
      throw new Error(error);
    }
    debug('Validated config.');
  }

  /**
   * Register the plugin with a provided set of events on a provided Hook system.
   *
   * @param {object} context - A Uttori-like context.
   * @param {object} context.hooks - An event system / hook system to use.
   * @param {Function} context.hooks.on - An event registration function.
   * @param {object} context.config - A provided configuration to use.
   * @param {object} context.config.events - An object whose keys correspong to methods, and contents are events to listen for.
   * @example <caption>AuthSimple.register(context)</caption>
   * const context = {
   *   hooks: {
   *     on: (event, callback) => { ... },
   *   },
   *   config: {
   *     [AuthSimple.configKey]: {
   *       ...,
   *       events: {
   *         bindRoutes: ['bind-routes'],
   *       },
   *     },
   *   },
   * };
   * AuthSimple.register(context);
   * @static
   */
  static register(context) {
    debug('register');
    if (!context || !context.hooks || typeof context.hooks.on !== 'function') {
      throw new Error("Missing event dispatcher in 'context.hooks.on(event, callback)' format.");
    }
    /** @type {AuthSimpleConfig} */
    const config = { ...AuthSimple.defaultConfig(), ...context.config[AuthSimple.configKey] };
    if (!config.events) {
      throw new Error("Missing events to listen to for in 'config.events'.");
    }
    for (const [method, events] of Object.entries(config.events)) {
      for (const event of events) {
        if (typeof AuthSimple[method] === 'function') {
          context.hooks.on(event, AuthSimple[method]);
        } else {
          debug(`Missing function "${method}" for key "${event}"`);
        }
      }
    }
  }

  /**
   * Add the login & logout routes to the server object.
   *
   * @param {object} server - An Express server instance.
   * @param {Function} server.post - Function to register route.
   * @param {object} context - A Uttori-like context.
   * @param {object} context.config - A provided configuration to use.
   * @example <caption>AuthSimple.bindRoutes(server, context)</caption>
   * const context = {
   *   hooks: {
   *     on: (event, callback) => { ... },
   *   },
   *   config: {
   *     [AuthSimple.configKey]: {
   *       loginPath: '/login',
   *       logoutPath: '/logout',
   *       loginMiddleware: [ ... ],
   *       logoutMiddleware: [ ... ],
   *     },
   *   },
   * };
   * AuthSimple.bindRoutes(server, context);
   * @static
   */
  static bindRoutes(server, context) {
    debug('bindRoutes');
    const { loginPath, logoutPath, loginMiddleware, logoutMiddleware } = { ...AuthSimple.defaultConfig(), ...context.config[AuthSimple.configKey] };
    debug('bindRoutes loginPath:', loginPath);
    debug('bindRoutes logoutPath:', logoutPath);
    server.post(loginPath, ...loginMiddleware, AuthSimple.login(context));
    server.post(logoutPath, ...logoutMiddleware, AuthSimple.logout(context));
  }

  /**
   * The Express route method to process the login request and provide a response or redirect.
   *
   * @param {object} context A Uttori-like context.
   * @param {object} context.config A provided configuration to use.
   * @returns {Function} - The function to pass to Express.
   * @example <caption>AuthSimple.login(context)(request, response, next)</caption>
   * server.post('/login', AuthSimple.login(context));
   * @static
   */
  static login(context) {
    return asyncHandler(async (request, response, next) => {
      debug('login');
      const { validateLogin, loginPath, loginRedirectPath } = { ...AuthSimple.defaultConfig(), ...context.config[AuthSimple.configKey] };

      if (typeof validateLogin !== 'function') {
        debug('Missing "validateLogin" function.');
        next();
        return;
      }

      // Validate login, returns profile.
      const profile = await validateLogin(request);
      debug('profile:', profile);
      if (profile) {
        // Set the response on the session.
        request.session.profile = profile;

        if (request.is('application/json')) {
          response.setHeader('Content-Type', 'application/json');
          response.json(profile);
        } else {
          response.status(302).redirect(loginRedirectPath);
        }
        return;
      }

      // Not logged in.
      if (request.is('application/json')) {
        response.setHeader('Content-Type', 'application/json');
        response.status(401).json({ error: true });
      } else {
        response.redirect(loginPath);
      }
    });
  }

  /**
   * The Express route method to process the logout request and clear the session.
   *
   * @param {object} context A Uttori-like context.
   * @param {object} context.config A provided configuration to use.
   * @returns {Function} - The function to pass to Express.
   * @example <caption>AuthSimple.login(context)(request, response, _next)</caption>
   * server.post('/logout', AuthSimple.login(context));
   * @static
   */
  static logout(context) {
    return asyncHandler((request, response, _next) => {
      debug('logout');
      const { logoutRedirectPath } = { ...AuthSimple.defaultConfig(), ...context.config[AuthSimple.configKey] };
      request.session.destroy((error) => {
        debug('Destroyed session:', error);
      });
      response.redirect(logoutRedirectPath);
    });
  }
}

module.exports = AuthSimple;
